### PR TITLE
[2.6][Debug] Fix fatal-errors handling on HHVM

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14967 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Fatal errors on HHVM use the 16777217 magic number.
But since `16777217 == E_ERROR + 0x1000000;` the only required tweak is to remove this leading `1`.

**Comment by staabm** [comment]
Feels like this should be reported as a php incompatibility to hhvm

File: src/Symfony/Component/Debug/Tests/ErrorHandlerTest.php [hunk 1]
```diff
@@ -393,6 +393,52 @@ public function testHandleFatalError()
...
+    {
+        try {
+            $handler = ErrorHandler::register();
+
+            $logger = $this->getMock('Psr\Log\LoggerInterface');
+            $logger
+                ->expects($this->once())
+                ->method('log')
+                ->with(
+                    $this->equalTo(LogLevel::EMERGENCY),
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/Tests/ErrorHandlerTest.php]: This has changed to LogLevel::CRITICAL in 2.7, be careful when merging

**Comment by nicolas-grekas** [comment]
I took some time to test the Debug component on HHVM.

The main differences are:
- the 5th $backtrace argument
- the hhvm.error_handling.call_user_handler_on_fatals=1 behavior
- the `16777217 == E_ERROR + 0x1000000` special error level

This PR handles all of them and gives full HHVM support to the component.
Ping @symfony/deciders

**Comment by digitalkaoz** [comment]
yeah :)

**Comment by stof** [comment]
:+1:

File: src/Symfony/Component/Debug/ErrorHandler.php [hunk 2]
```diff
@@ -402,10 +411,17 @@ public function handleError($type, $message, $file, $line, array $context)
             if ($this->scopedErrors & $type) {
                 $e['scope_vars'] = $context;
                 if ($trace) {
-                    $e['stack'] = debug_backtrace(true); // Provide object
+                    $e['stack'] = $backtrace ?: debug_backtrace(true); // Provide object
                 }
             } elseif ($trace) {
-                $e['stack'] = debug_backtrace(PHP_VERSION_ID >= 50306 ? DEBUG_BACKTRACE_IGNORE_ARGS : false);
+                if (null === $backtrace) {
+                    $e['stack'] = debug_backtrace(PHP_VERSION_ID >= 50306 ? DEBUG_BACKTRACE_IGNORE_ARGS : false);
```
Comment (Tobion) [inline src/Symfony/Component/Debug/ErrorHandler.php]: since we require php 5.3.9, this condition is irrelevant, isn't it?

Comment (stof) [inline src/Symfony/Component/Debug/ErrorHandler.php]: We don't require 5.3.9 in the 2.6 branch

Comment (Tobion) [inline src/Symfony/Component/Debug/ErrorHandler.php]: it should be changed in 2.7 then.

**Comment by Tobion** [comment]
:+1:

**Pull request merged**